denizen_scripts/survival/repo-link/items/system/soul_item_system.dsc

```diff
@@ -247,58 +337,78 @@ item_with_soul:
+      - define buffs_and_debuffs <[buffs_and_debuffs].with[debuffs].as[<list_single[<[debuffs]>]>]>
+    - define flavor "<&d>Soul Item<&nl><&e>Combine with Armor or Weapons"
+    - define NBT <[NBT].with[flavor].as[<[flavor]>].include[<[buffs_and_debuffs]>]>
+    - determine <proc[item_system_build_item].context[<item[soul].with[nbt=<[NBT]>]>]>
 
 item_with_soul_create:
   type: procedure
   definitions: item1|item2
   script:
     - if <[item1].material.name> == air || <[item2].material.name> == air:
       - determine gray_stained_glass_pane
-    - if !<list[<[item1]>|<[item2]>].filter[material.name.is[==].to[<script[soul].yaml_key[material]>]].get[1].has_nbt[soul]>:
+    - if !<list[<[item1]>].include[<[item2]>].filter[material.name.is[==].to[<script[soul].data_key[material]>]].first.has_nbt[soul]>:
       - determine gray_stained_glass_pane
-    - define soul_item <list[<[item1]>|<[item2]>].filter[has_nbt[soul]].get[1]>
-    - define soul_type <[soul_item].nbt[soul].before[,]>
-    - define soul_level <[soul_item].nbt[soul].after[,]>
-    - define rarity <[soul_item].nbt[rarity]>
-    - define item <list[<[item1]>|<[item2]>].filter[has_nbt[soul].not].get[1].material.name>
-    - define enchants <list[<[item1]>|<[item2]>].filter[has_nbt[soul].not].get[1].enchantments.with_levels.escaped||<list[]>>
-    - determine <proc[item_create_soul_item].context[<[item]>|<[soul_type]>|<[soul_level]>|<[rarity]>|<[enchants]>]>
+
+    - define soul_item <list[<[item1]>].include[<[item2]>].filter[has_nbt[soul]].first>
+    - define contexts <list.include[<list[<[item1]>|<[item2]>].filter[has_nbt[soul].not].first.material.name>]>
+    - define contexts <[contexts].include[<[soul_item].nbt[soul].get[type]>]>
+    - define contexts <[contexts].include[<[soul_item].nbt[soul].get[level]>]>
+    - define contexts <[contexts].include[<[soul_item].nbt[rarity]>]>
+    - define contexts <[context].include[<list_single[<list[<[item1]>|<[item2]>].parse_tag[<[parse_value].enchantments.with_levels>].combined.deduplicate>]>]>
+    - determine <proc[item_create_soul_item].context[<[Contexts]>]>
 
 item_create_soul_item:
   type: procedure
-  definitions: material|soul_type|level|rarity|enchants_escaped
+  definitions: material|soul_type|level|rarity|enchantments
   script:
-    - if <[material].matches[<script[item_system_global_data].yaml_key[regex_type_check.weapon]>]>:
-      - define base <script[item_system_global_data].yaml_key[defaults.damage.<[material]>]>
-    - foreach <script[item_system_global_data].yaml_key[soul_names.<[rarity]>.<[soul_type]>]> as:modifier:
-      - define stat <[modifier].before[/]>
-      - define type <[modifier].after[/]>
+    - if <[material].matches[<script[item_system_global_data].data_key[regex_type_check.weapon]>]>:
+      - define base <script[item_system_global_data].data_key[defaults.damage.<[material]>]>
+    - foreach <script[item_system_global_data].data_key[soul_names.<[rarity]>.<[soul_type]>]> key:stat as:type:
```

Comment:
initially, this looped across a hacked `key/value` list map, in the example for a soul item named "bear" with a rarity of "1":
`armor/buff|attack_speed/buff|health/debuff`
now loops across the mapping:
```yml
bear:
  health: buff
  melee_damage: buff
  attack_speed: debuff
```

Ideally, we should really remove regex, considering we have both:

A full list of all material names that we could check against the `item_system_global_data.nbt_slots` keys
Tags that can run this without `ElementTag.matches[]` such as `ElementTag.ends_with[]`